### PR TITLE
Block Library: Reimplement Reusable Block using EditorProvider for embedded post editor

### LINKED ISSUE #7119: Reusable Blocks: Reimplement reusable block as embedded editor
At the moment, fetching reusable blocks also parses their content and insert them to the `editor`'s block list state. This approach have some drawbacks:

 - We need to maintain a cycle dependency between `blocks` and `shared blocks`
 - Fetching shared blocks from the server retriggers the parsing which means if it's currently rendered, the block will unmount/remount.

I tried simplifying the state in #7080 but it fails at dealing with the nested blocks.

The idea of the PR is to avoid having to parse the reusable blocks (content property) when loading them, and when we insert a reusable block into the editor, don’t parse it and add the parsed blocks into the editor’s state because these parsed blocks are not part of the post content but just a property of the reusable block and is already present in the state in its serialized form.

So when we insert the reusable block, its edit method is responsible of parsing the content and updating it when the user saves the reusable block’s form.

But the problem happens when we have inner blocks because the way we render the `InnerBlocks` we assume these are present in state, we don’t pass the list of blocks to show explicitly to the `BlockList` as opposed to the `BlockEdit` (root) where we can just pass the attributes of the block to show.

The path forward is not clear:

- My “perfectionism” tells me we should refactor our component to avoid relying too much on state `getBlock( uid )` and accepts blocks as props.
- But this is a lot of work and we lose a lot of flexibility in using selectors etc...
- We can also try to load another Gutenberg instance :man-shrugging: when clicking the “edit” button of a reusable block, but this obviously requires a lot of work to make Gutenberg reusable this way (which should benefit things like multiple Gutenberg instances in a page) and this also means we’d need to figure out where to show the inspector of this inner instance, should it be shown above the outer instance’s inspector.
- Or we should keep our current hack: (load the shared blocks inner blocks into state, but try to improve it in a way that we don’t remount the shared blocks when we update them/trigger another fetch)

related #5010 #5228

### PR DESCRIPTION
**Pull request by aduth**
Closes #7119 
Previously: #7453

This pull request seeks to reimplement the reusable block as an embedded post editor, using `EditorProvider`.

**Work in Progress:**

This pull request currently cherry-picks a few related fixes (#14711), and includes changes which are likely more appropriate to be proposed in their own pull requests. It also lacks sufficient test coverage and documentation for newly proposed features.

Functionally, it largely works as expected.

I'm still trying to find the right balance of the extent to which this should be refactored. Currently, I'm considering to not remove any actions around fetching or saving reusable blocks as it impacts the inserter and block static/reusable conversion features. To that end, it would largely be a refactor primarily only impacting the block itself, but leaving all other behaviors intact.

### CODE REVIEW

**Comment by aduth** [comment]
I'm looking to refresh this soon, rebasing to resolve conflicts.

To note, the current commits b5c43e373461ad64b5b9d9809762be68edd017ba and 58c4a02d587b533e4933924fd4b6630fcb0dde95 have been resolved separately by #14711. These will be dropped from the branch.

I plan to open separate, individual pull requests for the commits 35b8e8e1c75b323051914b22f5cf76eb05c9043d and 53be3f2fc6bf95640b7ce9c94c7b055407a0154a, which serve standalone purpose (consistency with `BlockEditorProvider`, also related to #15949).

**Comment by aduth** [comment]
I've finished and pushed the rebase here. Functionally it seems to be in working order again. There are two failing unit test in `components/slot-fill/test/slot.js`, which I expect are related to more recent refactorings in #15541. I started down the path of refactoring the proposed Slot/Fill `withConsumerContext` of 71c3ec2107bd36a492d3e00a6a24d53e6d488b40 to a custom hook instead (similar to `@wordpress/data`'s [`useRegistry`](https://github.com/WordPress/gutenberg/blob/master/packages/data/src/components/registry-provider/use-registry.js)), but since this will require a refactor of [the `Slot` component](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/slot-fill/slot.js) to a function component (required for hooks), I will postpone this effort and propose the function refactoring separately (which should be done anyways).

**Comment by aduth** [comment]
> I plan to open separate, individual pull requests for the commits [35b8e8e](https://github.com/WordPress/gutenberg/commit/35b8e8e1c75b323051914b22f5cf76eb05c9043d) and [53be3f2](https://github.com/WordPress/gutenberg/commit/53be3f2fc6bf95640b7ce9c94c7b055407a0154a)

See #15989 and #15988 respectively.

File: packages/block-library/src/block/edit.js [hunk 1]
```diff
@@ -1,187 +1,175 @@
...
-import { BlockEdit } from '@wordpress/block-editor';
+import {
+	BlockList,
+	BlockControls,
+	BlockFormatControls,
+	FormatToolbar,
+	__experimentalBlockSettingsMenuFirstItem,
+	__experimentalBlockSettingsMenuPluginsExtension,
+} from '@wordpress/block-editor';
+import { EditorProvider } from '@wordpress/editor';
```
Comment (jorgefilipecosta) [inline packages/block-library/src/block/edit.js]: I would prefer if we could avoid the need to use the editor module in the reusable block. That would make us need to load the editor module in the widget screen, for example. It also makes our reusable block functionality very tied WordPress posts.
What if the BlockEditor settings accept a property that allows retrieving the blocks of the reusable block with a given id, and changing the blocks/title of a reusable block with a given id? The block would use these block editor settings to retrieve the information needed and would then instantiate a BlockEditor instead of an Editor.

This is not a blocker is just a future possibility.

File: packages/block-editor/src/components/rich-text/format-toolbar/index.js [hunk 2]
```diff
@@ -9,7 +9,9 @@ import { orderBy } from 'lodash';
  */
 
 import { __ } from '@wordpress/i18n';
-import { Toolbar, Slot, DropdownMenu } from '@wordpress/components';
+import { Toolbar, Slot, DropdownMenu, createSlotFill } from '@wordpress/components';
+
+const { Slot: ToolbarControlsSlot } = createSlotFill( 'RichText.ToolbarControls' );
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/rich-text/format-toolbar/index.js]: This change was only to have a reference to the slot so we can pass it to SlotFillProvider right?

File: packages/components/src/slot-fill/context.js [hunk 3]
```diff
@@ -21,22 +22,49 @@ class SlotFillProvider extends Component {
...
+	 */
+	proxy( functionName ) {
+		return ( name, ...args ) => {
+			const { slots: handledSlots } = this.props;
+
+			let handler = this[ functionName ];
+
+			if ( Array.isArray( handledSlots ) ) {
+				const isHandled = some( handledSlots, ( slotNameOrComponent ) => {
+					const { slotName = slotNameOrComponent } = slotNameOrComponent;
```
Comment (jorgefilipecosta) [inline packages/components/src/slot-fill/context.js]: This works well and is smart but makes the code a little hard to follow. Maybe we can just assign the name with const `slotName = ( typeof slotNameOrComponent === 'string' ) slotNameOrComponent : slotNameOrComponent.slotName;`
More verbose but I guess it is easier to follow.

File: packages/components/src/slot-fill/context.js [hunk 4]
```diff
@@ -21,22 +22,49 @@ class SlotFillProvider extends Component {
...
+						name.startsWith( slotName + '.' )
+					);
+				} );
+
+				if ( ! isHandled ) {
+					handler = this.props[ functionName ];
+				}
+			}
+
+			return handler.call( this, name, ...args );
```
Comment (jorgefilipecosta) [inline packages/components/src/slot-fill/context.js]: I'm probably missing something but is .call really required? Couldn't we just execute `return handler( name, ...args );`. Why do we have a need to change the "owner object"?

**Comment by jorgefilipecosta** [comment]
I spent some time testing this PR. Here are some regressions I found (nothing major):
- There is no way to select the reusable block title right after the creation. It always saves with the default title, without letting the user do any input after the creation. After it is saved the user can press the edit and edit the tile.
- If we select a block inside a reusable block the inspector still says the selected block is the reusable block. It correctly shows the settings of the selected block.
- If we multi-select blocks inside a reusable block, the inspector does not show a multi-select indication. I think we can solve this problem and the previous problem using a new slot that contains all the block inspector UI e.g: a slot where we render <BlockInspector />, similar to what I did in https://github.com/WordPress/gutenberg/pull/16203.
- It is not possible to insert a  reusable block inside another reusable block.
- If we have a block selected inside the reusable block editor and we insert a block using the global inserter the block gets inserted after the reusable block and not inside the reusable block. It is a complex to solve problem, the behavior in master is even worst -- nothing happens when the user tries to insert using the global inserter and has a block selected.


To note that even with this changes the reusable block still depends on the editor for the fetching of reusable blocks. But I guess it is something we can address after this PR.

File: packages/block-library/src/block/edit.js (hunk 1 above)
Comment (aduth) [reply packages/block-library/src/block/edit.js]: > What if the BlockEditor settings accept a property that allows retrieving the blocks of the reusable block with a given id, and changing the blocks/title of a reusable block with a given id? The block would use these block editor settings to retrieve the information needed and would then instantiate a BlockEditor instead of an Editor.

In what you're considering, where would you propose that the logic for saving / fetching reusable block as a post occur?

File: packages/block-editor/src/components/rich-text/format-toolbar/index.js (hunk 2 above)
Comment (aduth) [reply packages/block-editor/src/components/rich-text/format-toolbar/index.js]: > This change was only to have a reference to the slot so we can pass it to SlotFillProvider right?

Yes, rather than reference these by their string names (which so far we've considered an internal implementation detail), it seems we may instead want to reference the components themselves.

File: packages/components/src/slot-fill/context.js (hunk 3 above)
Comment (aduth) [reply packages/components/src/slot-fill/context.js]: > This works well and is smart but makes the code a little hard to follow.

Yeah, in retrospect I think you're right that I may have tried to be a little too clever here. Your simpler alternative seems to be an improvement.

File: packages/components/src/slot-fill/context.js (hunk 4 above)
Comment (aduth) [reply packages/components/src/slot-fill/context.js]: > I'm probably missing something but is .call really required? Couldn't we just execute `return handler( name, ...args );`. Why do we have a need to change the "owner object"?

I don't immediately recall why or if this was necessary. Looking at it now, it feels like it may have been at least necessary that a `this` context exist in the component instance function being called, but as I see it here, it's probably wrong to consider it the `this` of the _current_ instance (rather than the ancestor's), and I would think even without `.call`, the `this` would still be assigned in the ancestor (since most components define instance properties with `.bind` in the constructor).

**Comment by aduth** [comment]
> I spent some time testing this PR. Here are some regressions I found (nothing major):

Thanks for the detailed review! I'll look through these issues and see what can be done to improve them.

>To note that even with this changes the reusable block still depends on the editor for the fetching of reusable blocks. But I guess it is something we can address after this PR.

Yes, after a few failed attempts to switch everything over in the same pull request, I decided it would probably be best to do it in a few steps instead.

**Comment by aduth** [comment]
There is likely significant refactoring necessary to bring this up to date with respect to the introduction of "entity providers" (#17368).

**Comment by aduth** [comment]
This was largely superseded by #14367. There's still an open question whether Reusable Block should implement its own EditorProvider, vs. just the base BlockEditorProvider, considering that Reusable Block is implemented as saving to and from a post entity (if this is intended to be the abstraction that EditorProvider serves). In any case, between #14367 and https://github.com/WordPress/gutenberg/pull/14715#issuecomment-549418135, any work here will probably be better started from scratch. With that in mind, I'll close the pull request.
